README
yaffs2utils
-----------
The yaffs2utils is the collection of utilities to make/extract a YAFFS2/YAFFS1
image for Linux.

Currently, it inclueds the following tools:
(1) mkyaffs2: making a yaffs2 (yaffs1) image 
(2) unyaffs2: extract a yaffs2 (yaffs1) image made by mkyaffs2.
(3) unspare2: extract the spare(oob) layout from the NAND device.

This project is developed under the Debian Linux 5 (code Lenny). It can make a
yaffs2 image to be ran under my embedded system (Linux 2.6.20), and extract a
yaffs2 image made by itself. 

Due to some lack of the official utilities from yaffs2.net (A image made from
its "mkyaffs2image" tool cannot be ran normally under my embedded system because
of the wrong OOB origination), I re-wrote it almost and create the "mkyaffs2"
tool to make a yaffs2 image for Linux 2.6.20. A tool named "unyaffs2" is also
provided to extract the image made from its "mkyaffs2". The license for this 
package is stated as GPLv2 only.

Making a yaffs1 image to be ran properly is NOT verified, since I don't have an
embedded system device with the old NAND flash (512k page). I will be glad if
someone can help me to verify the integrity to make and extract a yaffs1 image.

* Project Page

	http://code.google.com/p/yaffs2utils/


HOWTO
-----
Building the source by "make", then enjoying them.


Usage
-----
* mkyaffs2

	./mkyaffs2 [-h] [-e] [-p pagesize] [-s sparesize] [-o oobimg] dirname imgfile

* unyaffs2

	./unyaffs2 [-h] [-e] [-p pagesize] [-s sparesize] [-o oobimg] [-f file] imgfile dirname

* unspare2

	./unspare2 [-h] [-e] devfile imgfile

After making a yaffs2 image, the tool "nandwrite" in the mtd-utils can be used
to recode data into the flash (with oob option), such as:

	nandwrite -a -o /dev/mtd${MTDNUM} ${YAFFS2IMAGE}

The option '-p' and '-s' can specify the page and spare size of the images, and
the information of the page and spare size should be available from your NAND
flash datasheet. If the oob layout is differed from the Linux default, the tool
"unspare2" can be used to extract the oob layout from the NAND device. If the
endian is different from your building system, The option "-e" is required while
executing the "unspare2" on the targets. Finally, by the option '-o' to specify
the oob image while using the "mkyaffs2/unyaffs2" can create/extract the YAFFS
image with the various oob layout for the different NAND devices.


Android
-------
Recently, I found that the most Android images are made from the official YAFFS2
tool - "mkyaffs2image". As the discussion in the previous section "yaffs2utils",
an image created by the official YAFFS2 tools (mkyaffs2image) can not be run on
the native Linux, because its oob layout is different from the default one in
the Linux MTD subsystem.

For an image is requested to be created/extracted with the compatibilities of
the one made by the official YFFS2 tool (mkyaffs2image), such as Android uses,
some methods are described:

1. Enable the CFLAGS in Makefile

   #CFLAGS	+= _HAVE_ANDROID


2. Modify the default the oob layout in the source by the following patch:

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
--- src.old/yaffs2utils_ecc.h	2011-06-20 16:23:45.000000000 +0800
+++ src.new/yaffs2utils_ecc.h	2011-06-20 16:22:55.000000000 +0800
@@ -32,7 +32,8 @@
         .eccpos = {40, 41, 42, 43, 44, 45, 46, 47,
                    48, 49, 50, 51, 52, 53, 54, 55,
                    56, 57, 58, 59, 60, 61, 62, 63},
-        .oobfree = {{.offset = 2, .length = 38}},
+//        .oobfree = {{.offset = 2, .length = 38}},
+        .oobfree = {{.offset = 0, .length = 38}},
 };
 
 static struct nand_ecclayout nand_oob_user = {0};
--------------------------------------------------------------------

3. Use "-o" option to specify the oob image, which can be available by using the
   tool "unspare2" on the target device, while creating/extracting the image.


LICENSE
-------
Except files under "src/yaffs2", yaffs2utils is published as GNU GPL version 2.
Files under "src/yaffs2" are branched from the official YAFFS2 and their license
are announced in their header.


TODO
----
1. Verify the implementation for yaffs1 images.
2. Extract the runtime image dumpped directly from the NAND flash.


NOTE
----
Files under "src/yaffs2" is branched from the official YAFFS2, because of the
compiling issues of the inconstant naming rules in the official YAFFS2 source.
These files are required by the current yaffs2utils. I will continue updating
these files untils the oficial source is integrated into the Linux kernel or it
has a stable release.


SPECIAL THANKS
--------------
1. Stijn Devriendt <HIGHGuY@gmail.com>


=============
Luen-Yung Lin
<penguin.lin@gmail.com>


CHANGES
2012-04-20	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.8
-------------

	* mkyaffs2:
	- Fix typos issue.
	- Add page size supports up to (16384) bytes.
	  - Since the latest Linux kernel (3.3) does not provide the default
	    oob scheme for page size > 4096 bytes, I wrote the "(8192), (16384)"
	    supports, instead of "8192, 16384", in the help message.

	* unyaffs2:
	- Fix typos issue.
	- Add page size supports up to (16384) bytes.
	  - Since the latest Linux kernel (3.3) does not provide the default
	    oob scheme for page size > 4096 bytes, I wrote the "(8192), (16384)"
	    supports, instead of "8192, 16384", in the help message.

	* others:
	- Update README for supportos of page and spare size.


2012-04-19	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.7
-------------

	* mkyaffs2:
	- Add a new option '-s' to specify the spare size.
	- Add page size support of 4096, (8192) bytes.

	* unyaffs2:
	- Add a new option '-s' to specify the spare size.
	- Add page size support of 4096, (8192) bytes.

	* others:
	- Fix #31: Doesn't compile properly on Mac OSX.


2012-04-11	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.6
-------------

	* unyaffs2:
	- Fix #30: unyaffs2: the chmod() should NOT apply on a symbol link.
	- Fix error displayed when image is NOT a regular file.

	* others:
	- Merge the latest YAFFS2 to 20120326 from the official repository.


2012-03-22	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.5
-------------

	* others:
	- Fix #26: Ordering of arguments prevents compile on certain systems.
	- Fix #27: make error:"mkyaffs2.c:19:21: -
		   error: configs.h:  No such file or directory.
	- Merge the latest YAFFS2 to 20120227 from the official release.


2012-01-19	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.4
-------------

	* unyaffs2
	- Fix #23: lchown after chmod destroies setuid/setgid.

	* others
	- Fix #22: Support unyaffs of yaffs2 img made from mkyaffs2image.
	  - Add a "Compatibility" section in the README.
	- Fix the "PATH_MAX" lack for OSX.
	  - Include "syslimits.h" for OSX.


2011-10-31	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.3
-------------

	* mkyaffs2
	- Fix typo.

	* unyaffs2
	- Fix typo.
	- Fix help messages.

	* others
	- Fix #21: makefile improvements, prints improvements, additional check.
	- Close #17: Can't compile on current Ubuntu (fixed in 0.2.2).
	- Close #20: Various defects (fixed in 0.2.2).


2011-09-13	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.2
-------------

	* mkyaffs2
	- Fix #17: Can't compile on current Ubuntu
	  - Replace "struct nand_ecclayout" with "nand_ecclayout_t".
	- Fix #20: Various defects.

	* unyaffs2 
	- Fix #17: Can't compile on current Ubuntu
	  - Replace "struct nand_ecclayout" with "nand_ecclayout_t".
	- Fix #19: Parsing OOB file fails when it is read-only.
	- Fix #20: Various defects.

	* unspare2
	- Fix #18: unspare2's -e parameter does not work.

	* others
	- Fix typo in "CHANGES".
	- Fix un-matched version string in "yaffs2utils.h".
	- Merge the latest YAFFS2 to 20110815 from the official release.
	- Remove the unused file "yaffs2utils_mtd.h"


2011-07-21	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.1
-------------

	* mkyaffs2
	- Fix #14: oobfree.
	  - The "mkyaffs2" will skip checking whether the spare size is enough.

	* unspare2
	- Enable unspare2 in Makefilie.
	- Replace "struct nand_ecclayout" with "nand_ecclayout_t".
	- Fix typo issues.

	* others
	- Use "mtd/mtd-user.h" instead of "yaffs2utils_mtd.h".


2011-07-07	Luen-Yung Lin <penguin.lin@gmail.com>

Version 0.2.0
-------------

	* mkyaffs2
	- Fix #4: Progress Bar Requirement.
	  - The 2-pass operation is applied to implment this feature. It will
	    scan the directory structure to build a filesystem tree first; then
	    making an image based on this tree.
	- Fix #7: Support far various OOB layout is lacking.
	  - An option "-o (oobimg)" is added to support the various OOB layout.
	    The OOB layout can be available by the tool "unspare2", and it will
	    be loaded into the specfied "nand_ecclayout" structure simply
	    without any verification.
	- Skip root object in yaffs2 image.
	  - From this version, the root object will NOT be written into the
	    yaffs2 image.
	- Add an option "-v (verbose)" to verbose details of making an image.
	  - By default, a progress bar will be displayed for the current status.
	    If the option "-v" is specified, a list of objects will be shown
	    instead of a progress bar.
	- Fix coding style typos.

	* unyaffs2
	- Fix #4: Progress Bar Requirement.
	  - As mkyaffs2, the 2-pass operation is appled. It will scan the full
	    image to build a filesystem tree based on their object id; and
	    extracting the image contents based on the tree. Finally, the files
	    attributes will also be restored.
	- Fix #7: Support far various OOB layout is lacking.
	  - An option "-o (oobimg)" is added to support the various OOB layout.
	    The OOB layout can be available by the tool "unspare2", and it will
	    be loaded into the specfied "nand_ecclayout" structure simply
	    without any verification.
	- Fix #8: Extract Only a Selection of Files.
	  - An option "-f (file)" is added to support extracting only the
	    selection of files. 
	- Skip root object in yaffs2 image.
	  - A root object (if it is existed) in the image will be skipped, 
	    and the pseudo root object will be used while extracting the image.
	- Add an option "-v (verbose)" to verbose details of making an image.
	  - By default, a progress bar will be displayed for the current status.
	    If the option "-v" is specified, a list of objects will be shown
	    instead of a progress bar.
	- Restore the attribute of files.
	- Fix coding style typos.

	* unspare2 (NEW)
	- A tool to dump the OOB layout from the NAND device.

	* others:
	- Update "README".
	- Use NAND ECC layout information from Linux kernel as default.
	- Merge the latest YAFFS2 to 20110628 from the official release.



src/yaffs2utils_ecc.h
/* 
 * yaffs2utils: Utilities to make/extract a YAFFS2/YAFFS1 image.
 * Copyright (C) 2010-2011 Luen-Yung Lin <penguin.lin@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef _YAFFS2UTILS_ECC_H_
#define _YAFFS2UTILS_ECC_H_

#ifndef _HAVE_BROKEN_MTD_H
 #include <mtd/mtd-user.h>
#else
 #include <yaffs2utils_mtd.h>
#endif

static nand_ecclayout_t nand_oob_16 = {
	.eccbytes	= 6,
	.eccpos		= {0, 1, 2, 3, 6, 7},
	.oobfree	= {{.offset = 8, .length = 8}},
};

static nand_ecclayout_t nand_oob_64 = {
	.eccbytes	= 24,
	.eccpos		= {40, 41, 42, 43, 44, 45, 46, 47,
			   48, 49, 50, 51, 52, 53, 54, 55,
			   56, 57, 58, 59, 60, 61, 62, 63},
#ifndef _HAVE_ANDROID
	.oobfree	= {{.offset = 2, .length = 38}},
#else
	.oobfree	= {{.offset = 0, .length = 38}},
#endif
};

static nand_ecclayout_t nand_oob_128 = {
	.eccbytes	= 48,
	.eccpos		= {
			   80, 81, 82, 83, 84, 85, 86, 87,
			   88, 89, 90, 91, 92, 93, 94, 95,
			   96, 97, 98, 99, 100, 101, 102, 103,
			   104, 105, 106, 107, 108, 109, 110, 111,
			   112, 113, 114, 115, 116, 117, 118, 119,
			   120, 121, 122, 123, 124, 125, 126, 127},
#ifndef _HAVE_ANDROID
	.oobfree	= {{.offset = 2, .length = 78}}
#else
	.oobfree	= {{.offset = 0, .length = 78}}
#endif
};

static nand_ecclayout_t nand_oob_user = {0};

#endif
